syntax = "proto3";

import "google/protobuf/empty.proto";
import "state.proto";

package tmapreduce;

option cc_generic_services = true;

message WorkerReplyMsg {
  bool ok = 1;
  string msg = 2;
  WorkerState state = 3;
}

message PrepareMsg {
  uint32 master_id = 1;
  string master_group = 2;
  string master_conf = 3;
  string job_type = 4;
}

message MapJobMsg {
  message MapKV{
    string key = 1;
    string value = 2;
  }
  uint32 master_id = 1;
  uint32 job_id = 2;
  uint32 subjob_id = 3;
  string name = 4;
  string type = 5;
  repeated MapKV map_kvs = 6;
}

message ReduceJobMsg {
  message ReduceKV{
    string key = 1;
    repeated string value = 2;
  }
  uint32 master_id = 1;
  uint32 job_id = 2;
  uint32 subjob_id = 3;
  string name = 4;
  string type = 5;
  repeated ReduceKV reduce_kvs = 6;
}


service WorkerService {
  rpc Beat(google.protobuf.Empty) returns (WorkerReplyMsg) {}
  rpc Prepare(PrepareMsg) returns (WorkerReplyMsg) {}
  rpc PrepareMap(MapJobMsg) returns (WorkerReplyMsg) {}
  rpc PrepareReduce(ReduceJobMsg) returns (WorkerReplyMsg) {}
  rpc Start(google.protobuf.Empty) returns (WorkerReplyMsg) {}
}